### PR TITLE
Making softmax numerically stable

### PR DESCRIPTION
**Pull request by joelshepherd**
Modified the softmax function to be numerically stable with large exponents. Method taken from [here](https://cs231n.github.io/linear-classify/#softmax-classifier).

I am fairly new to autodiff gradient functions, so my implementation of `amax` may be way off the mark (it certainly _looks_ wrong).

I originally wrote the below code based on the min/max gradient functions that already exist, but it would not converge my test model (where the current implementation does).
```ts
const mask = ctx.forward_inputs[0]
  .eq(ctx.forward_output)
  .astype(ctx.backward_input.dtype);
return ctx.backward_input.mul(mask)
```

### CODE REVIEW

**Comment by bwasti** [comment]
thanks! this is definitely a better approach.

With respect to gradient I think you need to mask out the non-max element indices (https://discuss.pytorch.org/t/differentiable-argmax/33020).
To help with debugging, you can add automated check for `amax` here: https://github.com/facebookresearch/shumai/blob/main/test/gradient.test.ts#L76 

As a reference, here's how the gradient state works: https://github.com/facebookresearch/shumai/pull/113#issuecomment-1302444022

**Comment by joelshepherd** [comment]
Thanks for the pointers! I solved what was missing from my earlier attempt with the help of those tests.

Side note: if in future we can index by tensor, a potentially more performant solution might be:
```ts
ctx.backwards_input.index(ctx.forward_input[0].argmax(...))
```

But for now, mask + sum seems to do the trick. This successfully converges my test models too.

**Comment by jacobkahn** [comment]
@bwasti -- index by tensor is supported in Flashlight, yep. Indexing by multiple tensors does an outer product index (rather than a Cartesian product index), although this is behavior we're looking to change.

Overall, indexing by a single tensor behave as you'd expect. cc @StrongerXi if I'm missing anything.

**Comment by StrongerXi** [comment]
> @bwasti -- index by tensor is supported in Flashlight, yep. Indexing by multiple tensors does an outer product index (rather than a Cartesian product index), although this is behavior we're looking to change.
> 
> Overall, indexing by a single tensor behave as you'd expect. cc @StrongerXi if I'm missing anything.

Ah one thing -- we don't have index by tensor support in OneDNN backend yet, because afaik OneDNN doesn't expose any primitives to help us with that.
